File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/actors/PlayActorTransferAdapter.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors;

import systems.zlink.framework.actors.ZLinkActorContext;
import systems.zlink.framework.actors.ZLinkActorTransferAdapter;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerInfo;

public final class PlayActorTransferAdapter
    implements ZLinkActorTransferAdapter<PlayActor> {
    @Override
    public java.util.concurrent.CompletionStage<ZLinkMessage> transferOut(PlayActor actor) {
        return java.util.concurrent.CompletableFuture.completedFuture(ZLinkMessage.of(new TransferState(
            actor.joinedRoomId(),
            actor.playerOrNull(),
            actor.destroyAfterEntrySpotJoin(),
            actor.disconnected())));
    }

    @Override
    public java.util.concurrent.CompletionStage<PlayActor> transferIn(
        String actorId,
        ZLinkActorContext context,
        ZLinkMessage state) {
        TransferState transferred = state.decode(TransferState.class);
        PlayActor actor = new PlayActor(actorId, context);
        if (transferred.player() != null) {
            actor.applyPlayer(transferred.player());
        }
        if (transferred.roomId() != null && !transferred.roomId().isBlank()) {
            actor.joinGame(transferred.roomId());
        }
        if (transferred.destroyAfterEntrySpotJoin()) {
            actor.markForDestroyAfterRoomLeave();
        }
        if (transferred.disconnected()) {
            actor.markDisconnected();
        }
        return java.util.concurrent.CompletableFuture.completedFuture(actor);
    }

    public record TransferState(
        String roomId,
        PlayerInfo player,
        boolean destroyAfterEntrySpotJoin,
        boolean disconnected) {
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/actors/PlayActorTransferAdapter.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors

import systems.zlink.framework.actors.ZLinkActorContext
import systems.zlink.framework.kotlin.ZLinkSuspendingActorTransferAdapter
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerInfo

class PlayActorTransferAdapter : ZLinkSuspendingActorTransferAdapter<PlayActor>() {
    override suspend fun transferOutSuspending(
        actor: PlayActor,
    ): ZLinkMessage = ZLinkMessage.of(
        TransferState(
            actor.joinedRoomIdOrNull(),
            actor.playerOrNull(),
            actor.destroyAfterEntrySpotJoin,
            actor.disconnected,
        ),
    )

    override suspend fun transferInSuspending(
        actorId: String,
        context: ZLinkActorContext,
        state: ZLinkMessage,
    ): PlayActor {
        val transferred = state.decode(TransferState::class.java)
        return PlayActor(actorId, context).also { actor ->
            transferred.player?.let(actor::applyPlayer)
            transferred.roomId?.takeIf(String::isNotBlank)?.let(actor::joinGame)
            if (transferred.destroyAfterEntrySpotJoin) actor.markForDestroyAfterRoomLeave()
            if (transferred.disconnected) actor.markDisconnected()
        }
    }

    data class TransferState(
        val roomId: String?,
        val player: PlayerInfo?,
        val destroyAfterEntrySpotJoin: Boolean,
        val disconnected: Boolean,
    )
}
